### DISCUSSION TITLE
How to run quantum-safe KEM algorithms?

### DISCUSSION THREAD

**Discussion by nikaijagganath** [open-quantum-safe/oqs-provider#380] [Q&A] [answered]
Hi there,

I am trying to use the quantum-safe KEM algorithms (specifically mlkem512) in a simple example to encapsulate/decapsulate a symmetric key, aes-256, that in turn will be used to encrypt/decrypt a text file. I am not sure how to run any of these KEM algorithms on the command line in my Ubuntu terminal. I was able to run a simple example of signing a text file with falcon512 but am getting errors when I try the KEM example.

**Code version/branch I am using:**
I managed to follow the instructions for cloning and installing https://github.com/open-quantum-safe/oqs-provider.git correctly. 

**$ openssl list -providers**
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.0.10
    status: active
  oqsprovider
    name: OpenSSL OQS Provider
    version: 0.5.4-dev
    status: active

**OpenSSL version running on my machine:**
**$ openssl version**
OpenSSL 3.0.10 1 Aug 2023 (Library: OpenSSL 3.0.10 1 Aug 2023)

**Attempted KEM commands with errors:**
**$openssl genpkey -algorithm mlkem512 -out sk.key**
Error writing key
40A72FFF707A0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):../crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
40A72FFF707A0000:error:04800073:PEM routines:do_pk8pkey:error converting private key:../crypto/pem/pem_pk8.c:133:

**$ openssl genpkey -aes-256-cbc -algorithm mlkem512 -out sk.pem**
Error writing key
40B75A1CEE7C0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):../crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
40B75A1CEE7C0000:error:04800073:PEM routines:do_pk8pkey:error converting private key:../crypto/pem/pem_pk8.c:133:

Please could you let me know what exactly I am doing incorrectly and how to run my KEM example. Thanks.

**Comment by ashman-p** [comment]
Hi, I wonder if you try specifying the provider in your command? 
openssl genpkey -algorithm mlkem512 -out sk.key -provider oqsprovider 
or
openssl genpkey -algorithm mlkem512 -out sk.key -provider oqsprovider -provider default

**Comment by baentsch** [comment]
Did you activate the KEM encoders in your build? [OQS_KEM_ENCODERS](https://github.com/open-quantum-safe/oqs-provider/blob/main/CONFIGURE.md#oqs_kem_encoders) by default is OFF. Check via `openssl list -encoders`.

**Comment by nikaijagganath** [comment]
@ashman-p: Thanks for the reply. This still doesn't work though; I think the issue is what @baentsch mentioned.

@baentsch: I think this is indeed the problem because when I run `openssl list -encoders`, none of the KEM algorithms come up. I made the mistake of thinking that running `fullbuild.sh` would build the complete software stack with all functionality operational. Please could you let me know how to go about activating these KEM encoders? Do I need to rebuild? Where exactly do I set these OQS_KEM_ENCODERS to "ON"? Thanks.

**Comment by baentsch** [comment]
> Where exactly do I set these OQS_KEM_ENCODERS to "ON"?

This is all documented in https://github.com/open-quantum-safe/oqs-provider/blob/main/CONFIGURE.md

May be easily set by convenience envvar  here https://github.com/open-quantum-safe/oqs-provider/blob/66ee77021b7ae35929e4659ac3f4b1e7642ddc67/scripts/fullbuild.sh#L7

**Reply by baentsch** [reply, marked as answer]
i.e., run `OQSPROV_CMAKE_PARAMS="-DOQS_KEM_ENCODERS=ON" ./scripts/fullbuild.sh`

**Reply by nikaijagganath** [reply]
Thank you, this fixed the issue!

**Comment by nikaijagganath** [comment]
Are there any other examples, apart from here https://github.com/open-quantum-safe/oqs-provider/blob/main/USAGE.md#activation, or other documentation on how to use openssl with these KEMs to encrypt an aes-256 key? I can't seem to get the encryption/decryption working.

**Reply by baentsch** [reply]
@nikaijagganath Please check out https://openquantumsafe.org/faq.html#whats-a-key-encapsulation-mechanism.

@dstebila It may be worth while adding an explicit section to the FAQ as to how (or not :) to encrypt using KEMs (or even provide guidance, as you once did somewhere -- just not remembering where; in any way surely not using `openssl` mechanisms).

**Reply by dstebila** [reply]
There was a discussion post about this in https://github.com/orgs/open-quantum-safe/discussions/1464.  I've added it to the FAQ in https://github.com/open-quantum-safe/www/pull/184
